WIP: rocAL - New Python Api

swetha097 opened the pull request.

LakshmiKumar23 commented on rocAL/rocAL_pybind/amd/rocAL/fn.py:
```diff
@@ -0,0 +1,534 @@
+    fog_image = b.Fog(Pipeline._current_pipeline._handle ,*(kwargs_pybind.values()))
+    return (fog_image)
+
+def brightness(*inputs, brightness=1.0, bytes_per_sample_hint=0, image_type=0,
```
Can you check on the fixed parameters - alpha and beta?

swetha097 replied:
I have added brightness_fixed augmentation, it uses alpha & beta params that can be exposed to the user

LakshmiKumar23 commented on rocAL/rocAL_pybind/amd/rocAL/fn.py:
```diff
@@ -0,0 +1,534 @@
+import rali_pybind as b
+from amd.rocAL.pipeline import Pipeline
+
+def blend(*inputs,**kwargs):
```
Can you check for the parameter - blend?

swetha097 replied:
Added necessary params for blend

LakshmiKumar23 commented on rocAL/rocAL_pybind/amd/rocAL/fn.py:
```diff
@@ -0,0 +1,534 @@
+    # pybind call arguments
+    kwargs_pybind = {"input_image0":inputs[0], "is_output":False ,"snpNoise": None}
+    snp_noise_added_image = b.SnPNoise(Pipeline._current_pipeline._handle ,*(kwargs_pybind.values()))
+    return (snp_noise_added_image)
```
File missing lens correction augmentation?

swetha097 replied:
Added lens correction augmentation

LakshmiKumar23 reviewed (changes requested):
@swetha097 Can you also check on rocal_pybind/run.sh. It queries for ```which python3.6``` specifically. But my system for instance has python3.8, and that script fails

rrawther commented on rocAL/rocAL/include/master_graph.h:
```diff
@@ -58,6 +58,8 @@ class MasterGraph
     Status copy_output(void* out_ptr, size_t out_size);
     Status copy_out_tensor_planar(void *out_ptr, RaliTensorFormat format, float multiplier0, float multiplier1, float multiplier2,
                     float offset0, float offset1, float offset2, bool reverse_channels, RaliTensorDataType output_data_type);
+    void *copy_hip_out_tensor(RaliTensorFormat format, float multiplier0, float multiplier1, float multiplier2,
```
Please keep the value return as status and pass a pointed to output as a parameter like the previous api

rrawther commented on rocAL/rocAL/include/master_graph.h:
```diff
@@ -88,6 +90,12 @@ class MasterGraph
     void create_randombboxcrop_reader(RandomBBoxCrop_MetaDataReaderType reader_type, RandomBBoxCrop_MetaDataType label_type, bool all_boxes_overlap, bool no_crop, FloatParam* aspect_ratio, bool has_shape, int crop_width, int crop_height, int num_attempts, FloatParam* scaling, int total_num_attempts, int64_t seed=0);
     const std::pair<ImageNameBatch,pMetaDataBatch>& meta_data();
     void set_loop(bool val) { _loop = val; }
+    void set_output_images(std::vector<Image*> output_images, unsigned int num_of_outputs)
```
Please pass vector by reference or pointer to avoid copy. Also if it is read only, please pass const reference

swetha097 replied:
Done

rrawther commented on rocAL/rocAL/rocAL_hip/rali_hip_kernels.cpp:
```diff
@@ -41,11 +41,11 @@ Hip_CopyInt8ToNHWC_fp32
     if ((x >= W) || (y >= H)) return;
     for (unsigned int n=0; n < nchw.x; n++) {
-        unsigned int srcIdx =  y*W + x;
+        unsigned int srcIdx =  y*W*3 + x*3;
```
modify it to srcIdx = dstIdx*3;

rrawther commented on rocAL/rocAL/source/master_graph.cpp:
```diff
@@ -316,6 +317,23 @@ MasterGraph::create_image(const ImageInfo &info, bool is_output)
 }
 
+void
+MasterGraph::set_output(Image* output_image)
```
I thought we have discussed to modify this into 2 separate routines

swetha097 replied:
We planned to make this change in the Tensor Branch Rajy

rrawther commented on rocAL/rocAL/source/master_graph.cpp:
```diff
@@ -1502,3 +1520,53 @@ MasterGraph::copy_out_tensor_planar(void *out_ptr, RaliTensorFormat format, floa
+#endif
+    _convert_time.end();
+    return static_cast<void *>(_output_tensor);
+}
```
add newline

swetha097 replied:
Done

Added a new line

rrawther commented on rocAL/rocAL/source/rali_api_augmentation.cpp:
```diff
@@ -1647,7 +1647,7 @@ RALI_API_CALL raliCropMirrorNormalize(RaliContext p_context, RaliImage p_input,
         ERR(e.what())
     }
-    return input; // Changed to input----------------IMPORTANT
+    return output;
```
why is this reverted back?

swetha097 replied:
This was a bug Rajy which was introduced long time back.

This did not affect any of our trainings because the Crop Mirror Normalizes output is never passed to any other node as an input

Hence reverted back the change

rrawther commented on rocAL/rocAL_pybind/amd/rocAL/fn.py:
```diff
@@ -0,0 +1,547 @@
+
+    return (random_bbox_crop,[],[],[])
+
+def bb_flip(*inputs, bytes_per_sample_hint = 0, horizontal = 1, ltrb = False, preserve =  False,seed = -1, vertical = 0, device = None):
```
please add comment specifying why this is a dummy node

swetha097 replied:
Done
Added the comment to tell why this is a dummy node

rrawther commented on rocAL/rocAL_pybind/amd/rocAL/fn.py:
```diff
@@ -0,0 +1,547 @@
+def nop(*inputs, device=None, preserve = False):
+    # pybind call arguments
+    kwargs_pybind = {"input_image0": inputs[0],"is_output": False }
+    nop_trandsformed_image = b.raliNop(Pipeline._current_pipeline._handle ,*(kwargs_pybind.values()))
```
better to name as nop_output (instead of other name). Applicable to other fns in the same file

LakshmiKumar23 replied:
please also check spelling for "transformed" if retaining name

swetha097 replied:
Done

rrawther commented on rocAL/rocAL_pybind/amd/rocAL/decoders.py:
```diff
@@ -0,0 +1,210 @@
+import amd.rocAL.types as types
```
I think, the rocAL name under amd should be all lowercase (rocal).

swetha097 replied:
Changed from rocAL to rocal

rrawther commented on rocAL/rocAL_pybind/amd/rocAL/pipeline.py:
```diff
@@ -240,6 +214,22 @@ def copyImage(self, array):
         b.raliCopyToOutput(
             self._handle, np.ascontiguousarray(out, dtype=array.dtype))
 
+    def copyToHipTensorNHWC(self, multiplier, offset, reverse_channels, tensor_dtype):
```
better to rename to copyToGpu..() instead of using Hip

swetha097 replied:
Renamed

rrawther commented on rocAL/rocAL_pybind/amd/rocAL/plugin/pytorch.py:
```diff
@@ -29,14 +29,14 @@ def __next__(self):
             for i in range(self.bs):
                 size = b.getImageNameLen(self.loader._handle,i)
                 print(size)
                 self.array = np.array(["                 "])
```
This is not a good way to define array with spaces??

swetha097 replied:
Changed it

rrawther commented on rocAL/rocAL_pybind/amd/rocAL/readers.py:
```diff
@@ -0,0 +1,101 @@
+                 save_img_ids=False, seed=-1, shard_id=0, shuffle_after_epoch=False, size_threshold=0.1,
+                 skip_cached_images=False, skip_empty=False, stick_to_shard=False, tensor_init_bytes=1048576):
+
+    Pipeline._current_pipeline._reader = "COCOReader"
```
seems like shard_id is passed to fn and not used?

swetha097 replied:
Yes, Rajy 
This change was done to match the outer API with arguments same as DALI's .

rrawther commented on rocAL/rocAL_pybind/example/rali-python-unittest.py:
```diff
@@ -122,9 +117,9 @@ def define_graph(self):
-                
-                
-                
+
```
remove empty lines

swetha097 replied:
Resolved

rrawther commented on rocAL/rocAL_pybind/example/tf_classification_pipeline.py:
```diff
@@ -72,15 +72,15 @@ def main():
 	imageIterator.reset()
-			
-		
+
```
remove empty lines

swetha097 replied:
Done

rrawther reviewed (changes requested):
address review comments

rrawther commented:
@swetha097 : Can you please let me know if you have resolved the review comments in this PR?

rrawther commented on rocAL/rocAL_pybind/amd/rocAL/fn.py:
```diff
@@ -128,7 +128,7 @@ def gamma_correction(*inputs, gamma=0.5, device=None):
     return (gamma_correction_image)
 
-def hue(*inputs, bytes_per_sample_hint=0,  hue=0.0, image_type=0,
+def hue(*inputs, bytes_per_sample_hint=0,  hue=None, image_type=0,
```
Why the hue=None here?

swetha097 replied:
That is a default value Rajy
Later a FloatParam is created if a float value is passed by the user
else, a None would passed to the C API call indicating the default value will be picked
hue = b.CreateFloatParameter(hue) if isinstance(hue, float) else hue

swetha097 commented:
> @swetha097 : Can you please let me know if you have resolved the review comments in this PR?

@rrawther : yes Rajy, all comments have been resolved except one comment

> 
I thought we have discussed to modify this into 2 separate routines
This change will be added in Tensor branch with new API

LakshmiKumar23 commented:
> @swetha097 Can you also check on rocal_pybind/run.sh. It queries for `which python3.6` specifically. But my system for instance has python3.8, and that script fails

@swetha097 Have you updated this as well?

LakshmiKumar23 commented:
@swetha097 Seems like the fn.py file got deleted? Can you check on that?

swetha097 commented:
> > @swetha097 Can you also check on rocal_pybind/run.sh. It queries for `which python3.6` specifically. But my system for instance has python3.8, and that script fails
> 
> @swetha097 Have you updated this as well?

No Lakshmi, I haven't updated this. I tried changing all the instances of 3.6 to 3.8 in run.sh file , that works.
But I need to make this change a generic one.

swetha097 commented:
> @swetha097 Seems like the fn.py file got deleted? Can you check on that?

@LakshmiKumar23 I will check on this

swetha097 commented:
> > @swetha097 Seems like the fn.py file got deleted? Can you check on that?
> 
> @LakshmiKumar23 I will check on this

@Lakshmi, The new rocal folder was missed in the previous commit.
Added it .

LakshmiKumar23 commented:
> > > @swetha097 Can you also check on rocal_pybind/run.sh. It queries for `which python3.6` specifically. But my system for instance has python3.8, and that script fails
> > 
> > 
> > @swetha097 Have you updated this as well?
> 
> No Lakshmi, I haven't updated this. I tried changing all the instances of 3.6 to 3.8 in run.sh file , that works. But I need to make this change a generic one.

@swetha097 Yes I tried this too and it worked when I changed it to python3.8 on my dev system

swetha097 commented:
> 
@LakshmiKumar23 , I have made rocal_pybind/run.sh generic which would work for any python3 version

rrawther commented on rocAL/rocAL/include/randombboxcrop_meta_data_reader.h:
```diff
@@ -87,7 +87,7 @@ class RandomBBoxCrop_MetaDataReader
     virtual std::vector<std::vector <float>>  get_batch_crop_coords(const std::vector<std::string>& image_names) = 0; // returns the crop coords for a batch
     virtual void release() = 0; // Deletes the loaded information
-    virtual void set_meta_data(std::shared_ptr<MetaDataReader> meta_data_reader) = 0;
+    virtual void set_meta_data(std::shared_ptr<MetaDataReader> meta_data_reader,MetaDataReaderType config_type) = 0;
```
adjust space between arguments

swetha097 replied:
Done

rrawther commented on rocAL/rocAL/source/randombboxcrop_reader.cpp:
```diff
@@ -45,9 +45,35 @@ void RandomBBoxCropReader::init(const RandomBBoxCrop_MetaDataConfig &cfg)
 {
-    _meta_data_reader = std::static_pointer_cast<COCOMetaDataReader>(meta_data_reader);
+
+    switch(type) {
```
it shouldn't check the reader type here. Instead access the function through base class

swetha097 replied:
Made changes, testing it with SSD training

swetha097 replied:
Tested, SSD Training converged

rrawther commented on rocAL/rocAL_pybind/amd/rocal/decoders.py:
```diff
@@ -0,0 +1,244 @@
+            "shuffle": random_shuffle,
+            "loop": False,
+            "decode_size_policy": types.USER_GIVEN_SIZE,
+            "max_width": 2000,
```
Looks like only TFRecordReaderDetection uses user_size?

swetha097 replied:
Yes Rajy

rrawther replied:
Do you know why?

rrawther commented on rocAL/rocAL_pybind/amd/rocal/fn.py:
```diff
@@ -0,0 +1,562 @@
+
```
please remove all unused imports and fix Codacy errors

swetha097 replied:
The fn.py requires the readers , decoders to be imported from readers.py / decoders.py so that it can be accessed as
fn.readers.function() in example scripts

But we don't require to use the readers / decoders inside the fn.py file 
Its for accessing similar to DALI

rrawther commented on rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py:
```diff
@@ -1,7 +1,7 @@
 import torch
 import numpy as np
 import rali_pybind as b
```
remove empty file above __init__.y

swetha097 replied:
The __init__.py files are required to make Python treat the directories as containing packages
In the simple case, they can also be empty.
[__init__.py](http://docs.python.org/tutorial/modules.html#packages):

rrawther commented on rocAL/rocAL_pybind/amd/rocal/types.py:
```diff
@@ -34,7 +34,9 @@
 from rali_pybind.types import NCHW
 
-
+#     RaliDecodeDevice
```
is the empty file above indented? Please remove

swetha097 replied:
Yes Rajy
The init.py files are required to make Python treat the directories as containing packages
In the simple case, they can also be empty.
[init.py files can be empty](http://docs.python.org/tutorial/modules.html#packages)

rrawther commented on rocAL/rocAL_pybind/example/new_api/rocAL_api_coco_pipeline_test.py:
```diff
@@ -0,0 +1,281 @@
+                else:
+                    img = torch.from_numpy(self.out)
+                    for idx in range(self.n):
+                        draw_all_images(img[i],self.image_id[i],idx, self.path)
```
The definition of function taking only 3 arguments. Has this been tested?

swetha097 replied:
There were 2 functions for the same functionality, removed one

rrawther commented on rocAL/rocAL_pybind/example/new_api/rocAL_api_coco_pipeline_test.py:
```diff
@@ -0,0 +1,281 @@
+
+if __name__ == '__main__':
+    main()
+
```
remove emply lines at the end

rrawther commented on rocAL/rocAL_pybind/example/new_api/rocAL_api_python_unittest.py:
```diff
@@ -0,0 +1,183 @@
+        #                                 mean=[0,0,0],
+        #                                 std=[1,1,1])
+
+        if augmentation_num == 0 or augmentation_num == 1:
```
please change augmentation_num to string (name so it is more clear)

swetha097 replied:
I was working on modifying the test suite for python. Added the above change in latest commit

rrawther commented on rocAL/rocAL_pybind/example/new_api/rocAL_api_test.py:
```diff
@@ -0,0 +1,73 @@
+    world_size = 1
+    rali_cpu= True
+    rali_device = 'cpu' if rali_cpu else 'gpu'
+    decoder_device = 'cpu' if rali_cpu else 'mixed'
```
we don't support mixed devices in rocal and so as padding. Please remove those

swetha097 replied:
Removed in the latest commit

rrawther commented on rocAL/rocAL_pybind/example/new_api/rocAL_api_test_file.py:
```diff
@@ -0,0 +1,194 @@
+
+if __name__ == '__main__':
+    main()
+
```
remove empty lines at the end

swetha097 replied:
Removed in latest commit

rrawther commented on rocAL/rocAL_pybind/rali_pybind.cpp:
```diff
@@ -235,6 +235,10 @@ namespace rali{
             .value("NHWC",RALI_NHWC)
```
looks like the above examples for training was deleted without converting to new API. We need to have them ported as well

rrawther replied:
the file needs to be renamed to rocal_pybind.cpp

swetha097 replied:
Have added this change as a part of rocal restructuring development branch of ours

rrawther reviewed (commented):
please address review comments

swetha097 commented on rocAL/rocAL/include/randombboxcrop_reader.h:
```diff
@@ -78,14 +81,14 @@ class RandomBBoxCropReader: public RandomBBoxCrop_MetaDataReader
     CropCordBatch * get_output() override { return _output; }
     bool is_entire_iou(){return _entire_iou;}
-    void set_meta_data(std::shared_ptr<MetaDataReader> meta_data_reader) override;
+    void set_meta_data(std::shared_ptr<MetaDataReader> meta_data_reader,MetaDataReaderType config_type ) override;
```
Done

swetha097 commented on rocAL/rocAL/source/meta_node_crop.cpp:
```diff
@@ -35,13 +35,15 @@ void CropMetaNode::update_parameters(MetaDataBatch* input_meta_data)
     }
-    _meta_crop_param = _node->get_crop_param();    
+    _meta_crop_param = _node->get_crop_param();
     _dst_width = _node->get_dst_width();
```
Done

swetha097 commented on rocAL/rocAL/source/master_graph.cpp:
```diff
@@ -1231,7 +1253,7 @@ void MasterGraph::create_randombboxcrop_reader(RandomBBoxCrop_MetaDataReaderType
     RandomBBoxCrop_MetaDataConfig config(label_type, reader_type, all_boxes_overlap, no_crop, aspect_ratio, has_shape, crop_width, crop_height, num_attempts, scaling, total_num_attempts, seed);
     _randombboxcrop_meta_data_reader = create_meta_data_reader(config);
-    _randombboxcrop_meta_data_reader->set_meta_data(_meta_data_reader);
+    _randombboxcrop_meta_data_reader->set_meta_data(_meta_data_reader, _reader_config);
```
Removed in the latest commit

swetha097 commented on rocAL/rocAL/source/image_loader.cpp:
```diff
@@ -50,6 +50,7 @@ ImageLoader::~ImageLoader()
 
 void ImageLoader::shut_down()
 {
+    stop_internal_thread();
```
Done

swetha097 commented on rocAL/rocAL/source/master_graph.cpp:
```diff
@@ -1139,6 +1157,7 @@ MetaDataBatch * MasterGraph::create_coco_meta_data_reader(const char *source_pat
     if( _meta_data_reader)
         THROW("A metadata reader has already been created")
     MetaDataConfig config(label_type, reader_type, source_path, std::map<std::string, std::string>(), std::string());
+    _reader_config = config.reader_type();
```
Made changes in the recent commits

swetha097 commented on rocAL/rocAL/source/caffe2_meta_data_reader.cpp:
```diff
@@ -53,7 +53,7 @@ void Caffe2MetaDataReader::add(std::string _image_name, int label)
         return;
     }
-    _map_content.insert(pair<std::string, std::shared_ptr<Label>>(_image_name, info));
+    _map_content.insert(pair<std::string, std::shared_ptr<MetaData>>(_image_name, info));
```
We had changed this to use randombbox crop map content for all the detection based readers instead of just coco reader as in previous case

swetha097 commented on rocAL/rocAL/rocAL_hip/rali_hip_kernels.cpp:
```diff
@@ -46,15 +46,21 @@ Hip_CopyInt8ToNHWC_fp32
     for (unsigned int n=0; n < nchw.x; n++) {
         unsigned int srcIdx =  (y*W + x) * C;     // src is RGB
-        unsigned int dstIdx =  y*W + x;
+        unsigned int dstIdx = ( y*W + x)*C;
```
The implementation of NHWC hip kernel was exactly same as NCHW. The output of NHWC kernel when dumping images was wrong. Changed it to correct the implementation

rrawther commented on rocAL/rocAL/source/rali_api_meta_data.cpp:
```diff
@@ -236,22 +236,25 @@ RALI_API_CALL raliGetImageLabels(RaliContext p_context, int* buf)
-    if(!meta_data.second)
-        THROW("No label has been loaded for this output image")
+    return context->master_graph->bounding_box_batch_count(buf, meta_data.second);
+#if 0
```
please remove this commented block. I forgot that in my branch

rrawther commented on rocAL/rocAL_pybind/amd/rocal/pipeline.py:
```diff
@@ -256,13 +246,44 @@ def encode(self, bboxes_in, labels_in):
+    def set_seed(self,seed=0):
+        return b.setSeed(seed)
+
+    def create_int_param(self,value):
```
I think it is complaining because it doesn't need to be a member of class in this case. Does this need to be a classmember? Please follow the link below to specify the function decorator type
https://docs.quantifiedcode.com/python-anti-patterns/correctness/method_could_be_a_function.html

swetha097 replied:
Yes, this needs to class members since we use these functions in the augmentation app access via pipeline method

rrawther commented on rocAL/rocAL_pybind/example/new_api/rocAL_api_coco_pipeline.py:
```diff
@@ -0,0 +1,272 @@
+    def __init__(self, pipelines, tensor_layout=types.NCHW, reverse_channels=False, multiplier=None, offset=None, tensor_dtype=types.FLOAT, device="cpu", display=False):
+
+        # self._num_gpus = len(pipelines)
+        assert pipelines is not None, "Number of provided pipelines has to be at least 1"
```
the Codacy error is because assert is used without try except. Please change it

swetha097 replied:
Changes done in latest commit

rrawther commented on rocAL/rocAL_pybind/example/new_api/rocAL_api_coco_pipeline.py:
```diff
@@ -0,0 +1,272 @@
+
+if __name__ == '__main__':
+    main()
+
```
remove empty lines

swetha097 replied:
Changes done in latest commit

rrawther reviewed (changes requested):
@swetha097 : Still there are lots of Codacy warnings other review comments which are not resolved. We can have a short call to resolve these

swetha097 commented:
> @swetha097 : Still there are lots of Codacy warnings other review comments which are not resolved. We can have a short call to resolve these

@rrawther : I was working on modifying the test suite for python. 
We can now test all the reader pipeline's from python at one go

- Added the argument parser in parse_config.py
- 
- README.md has also been added (We can either test each file singularly & at one shot )
- 
- This should help us in testing
- 
- I have also added a mini database containing smaller Image datasets (jpeg, coco, caffe , caffe2, tf) ->  rocAL_pybind/example/new_api/mini_db

rrawther reviewed (changes requested):
@swetha097 : We don't want to include all the data for testing as part of this PR. This has to be a separate repo with optional download. Please remove all the test data files from the PR

LakshmiKumar23 commented on rocAL/rocAL_pybind/example/new_api/READERS_TEST_FILE.sh:
```diff
@@ -0,0 +1,251 @@
+if [[ rocAL_api_coco_pipeline -eq 1 ]]; then
+
+    # Mention dataset_path
+    data_dir=$ROCAL_DATA_PATH/coco/coco_10_img/val_10images_2017/
```
@swetha097 The path is missing "/rocal_data/". Causes an error

swetha097 replied:
@LakshmiKumar23 , we need to export ROCAL_DATA_PATH=/Absolute/Path/Of/rocal_data/
[](https://github.com/swetha097/MIVisionX-data/tree/rocal_data_samples/rocal_data)

LakshmiKumar23 commented on rocAL/rocAL_pybind/example/new_api/READERS_TEST_FILE.sh:
```diff
@@ -0,0 +1,251 @@
+
+    # Mention dataset_path
+    # Detection
+    data_dir=$ROCAL_DATA_PATH/caffe/caffe_detection/lmdb_record/
```
@swetha097 Same missing "/rocal_data/"

swetha097 replied:
we need to export ROCAL_DATA_PATH=/Absolute/Path/Of/rocal_data/

We can see the dumped image outputs in OUTPUTS_IMAGES_PYTHON/NEW_API/

LakshmiKumar23 commented on rocAL/rocAL_pybind/example/new_api/READERS_TEST_FILE.sh:
```diff
@@ -0,0 +1,251 @@
+
+    # Mention dataset_path
+    # Classification
+    data_dir=$ROCAL_DATA_PATH/caffe2/classfication/imagenet_val5_encode/
```
@swetha097 same as above

LakshmiKumar23 commented on rocAL/rocAL_pybind/example/new_api/READERS_TEST_FILE.sh:
```diff
@@ -0,0 +1,251 @@
+
+    # Mention dataset_path
+    # Detection
+    data_dir=$ROCAL_DATA_PATH/caffe2/detection/lmdb_records/
```
@swetha097 same as above

LakshmiKumar23 commented on rocAL/rocAL_pybind/example/new_api/PYTHON_UNITTEST_TEST_FILE.sh:
```diff
@@ -0,0 +1,63 @@
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name saturation --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.saturation.${CURRENTDATE}
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name color_twist --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.color_twist.${CURRENTDATE}
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name crop_mirror_normalize --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.crop_mirror_normalize.${CURRENTDATE}
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name nop --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.nop.${CURRENTDATE}
```
@swetha097  nop fails with the following error
```
OK: loaded 80 kernels from libvx_rpp.so
[ERR]  {raliVerify}   { verify } vxVerifyGraph failed -1
[Errno 17] File exists: 'OUTPUT_IMAGES_PYTHON/NEW_API/FILE_READER/nop'
Pipeline has been created succesfully
Verify graph failed
```

LakshmiKumar23 commented on rocAL/rocAL_pybind/example/new_api/PYTHON_UNITTEST_TEST_FILE.sh:
```diff
@@ -0,0 +1,63 @@
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name copy --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.copy.${CURRENTDATE}
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name rotate_fisheye_fog --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.rotate_fisheye_fog.${CURRENTDATE}
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name resize_brightness_jitter --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.resize_brightness_jitter.${CURRENTDATE}
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name vignetter_blur --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.vignetter_blur.${CURRENTDATE}
```
@swetha097 All values of tensor = 0. Not sure if this is the expected output for vignette_blur

LakshmiKumar23 commented on rocAL/rocAL_pybind/example/new_api/PYTHON_UNITTEST_TEST_FILE.sh:
```diff
@@ -0,0 +1,63 @@
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name hue_rotate_blend --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.hue_rotate_blend.${CURRENTDATE}
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name warp_affine --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.warp_affine.${CURRENTDATE}
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name fish_eye --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.fish_eye.${CURRENTDATE}
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name vignette --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.vignette.${CURRENTDATE}
```
@swetha097  All values for tensor = 0. Not sure if this is the expected result

LakshmiKumar23 commented on rocAL/rocAL_pybind/example/new_api/PYTHON_UNITTEST_TEST_FILE.sh:
```diff
@@ -0,0 +1,63 @@
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name blur --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.blur.${CURRENTDATE}
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name hue_rotate_blend --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.hue_rotate_blend.${CURRENTDATE}
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name warp_affine --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.warp_affine.${CURRENTDATE}
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name fish_eye --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 2>&1 | tee -a run.log.fish_eye.${CURRENTDATE}
```
@swetha097 All values for tensor = 0. Not sure if this is the expected result

rrawther commented on rocAL/rocAL_pybind/example/new_api/rocAL_api_video_pipeline.py:
```diff
@@ -0,0 +1,155 @@
+
+if __name__ == '__main__':
+    main()
+
```
please remove empty lines here

rrawther commented on rocAL/rocAL_pybind/example/new_api/rocAL_api_video_pipeline.py:
```diff
@@ -0,0 +1,155 @@
+        self.reverse_channels = reverse_channels
+        self.tensor_dtype = tensor_dtype
+        self.batch_size = self.loader._batch_size
+        self.w = self.loader.getOutputWidth()
```
these are image specific parameters. Can you specify video sequence length and batch_size etc.. The sample has to be clear that it is iterating video files

rrawther commented on rocAL/rocAL_pybind/README.md:
```diff
@@ -1,18 +1,18 @@
 * [rocAL C/C++ Library](../rocAL#prerequisites)
 * CMake Version 3.10 or higher
-* Python 3.6 
+* Python 3.6
```
please mention python3.6 or higher since newer OS has python3.7

swetha097 replied:
Done

rrawther reviewed (changes requested):
@LakshmiKumar23 to make sure all tests pass

LakshmiKumar23 commented on rocAL/rocAL_pybind/example/new_api/READERS_TEST_FILE.sh:
```diff
@@ -0,0 +1,296 @@
+if [[ rocAL_api_python_unittest -eq 1 ]]; then
+
+    # Mention dataset_path
+    data_dir=$ROCAL_DATA_PATH/rocal_data/images_jpg/labels_folder/
```
@swetha097 This is what I am referring to. Line 103 vs line 78

rrawther commented on rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py:
```diff
@@ -50,25 +49,50 @@ def __iter__(self):
+                elif self.tensor_dtype == types.FLOAT16:
+                    self.out = torch.empty((self.bs*self.n, int(self.h/self.bs), self.w, self.p), dtype=torch.float16, device=torch_gpu_device)
         if(self.loader._oneHotEncoding == True):
             self.labels = np.zeros((self.bs)*(self.loader._numOfClasses),dtype = "int32")
```
need to allocate labels tensor also in host or device.  It is better to use torch.zeros with device attribute for labels tensor

rrawther commented on rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py:
```diff
@@ -144,24 +167,22 @@ def __next__(self):
 
         else:
             if(self.loader._oneHotEncoding == True):
                 self.loader.GetOneHotEncodedLabels(self.labels)
```
please pass device type (host or gpu) for GetOneHotEncodedLabels, so that it can copy to the correct destination. In case of HIP backend, please copy from internal rocal cpu memory to HIP mem pointer of the labels tensor, so we can avoid all .cuda() calls in the training script

swetha097 replied:
Done

rrawther commented on rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py:
```diff
@@ -144,24 +167,22 @@ def __next__(self):
+                    for i in range(self.bs):
+                        img = (self.out)
+                        draw_patches(img[i], i, 0)
                 self.loader.getImageLabels(self.labels)
```
please allocate labels as GPU tensor for device and copy to device memory directly in getImageLables rocal c++ call

swetha097 replied:
Done & tested the example file

rrawther commented on rocAL/rocAL_pybind/amd/rocal/random.py:
```diff
@@ -0,0 +1,12 @@
+    output_array = b.CreateIntRand(values, frequencies)
+    return output_array
+
+def uniform(*inputs,range=[-1, 1], device=None):
```
change it to another name (may be rng) so Codacy won't complain

rrawther commented on rocAL/rocAL_pybind/amd/rocal/readers.py:
```diff
@@ -0,0 +1,140 @@
+    bboxes = []
+    kwargs_pybind = {"source_path": annotations_file, "is_output":True}
+    b.setSeed(seed)
+    meta_data = b.COCOReader(Pipeline._current_pipeline._handle ,*(kwargs_pybind.values()))
```
looks like empty labels and boxes are returned from here which can cause problem in running the pipeline

swetha097 replied:
Rajy, This was done to Match DALI's API

rrawther commented on rocAL/rocAL_pybind/example/new_api/tf_petsTrainingExample/train_withRALI_withTFRecordReader.py:
```diff
@@ -135,13 +102,59 @@ def main():
+		resized = fn.resize(images, resize_x=crop_size[0], resize_y=crop_size[1])
+		flip_coin = fn.random.coin_flip(probability=0.5)
+		cmn_images = fn.crop_mirror_normalize(resized,
+													crop=(crop_size[1], crop_size[0]),
```
please adjust these spacing to be consistant

rrawther reviewed (changes requested):
please address requested review comments.
@LakshmiKumar23 : please test all the python tests and training scripts

LakshmiKumar23 commented:
@rrawther @kiritigowda @swetha097 All tests are passing with this PR. Can be merged.

rrawther commented on rocAL/rocAL/source/rali_api_meta_data.cpp:
```diff
@@ -216,9 +216,9 @@ RALI_API_CALL raliGetImageId(RaliContext p_context,  int* buf)
 {
-    
+
     if (!p_context)
```
check if buf is valid (not null) and throw error if so. Applies to all the functions below

rrawther commented on rocAL/rocAL_pybind/amd/rocal/fn.py:
```diff
@@ -0,0 +1,553 @@
+
+from amd.rocal import readers
+from amd.rocal import decoders
+from amd.rocal import random
```
please remove unused import

rrawther commented on rocAL/rocAL_pybind/amd/rocal/pipeline.py:
```diff
@@ -120,6 +122,12 @@ def __init__(self, batch_size=-1, num_threads=-1, device_id=-1, seed=-1,
         self._castLabels = False
+        self._prev_input_image = None
+        self._current_output_image = None
+        self._current_pipeline
```
what is this supposed to mean?

rrawther commented on rocAL/rocAL_pybind/amd/rocal/pipeline.py:
```diff
@@ -231,8 +203,10 @@ def define_graph(self):
         """This function is defined by the user to construct the
         graph of operations for their pipeline.
         It returns a list of outputs created by calling RALI Operators."""
+        print("define_graph NotImplemented")
```
modify the comment to say definegraph is deprecated

rrawther commented on rocAL/rocAL_pybind/amd/rocal/pipeline.py:
```diff
@@ -241,28 +215,81 @@ def copyImage(self, array):
+    def __exit__(self, exception_type, exception_value, traceback):
+        pass
+
+    def set_seed(self,seed=0):
```
this seems to be incorrect. set_seed() is not global. It should set_seed() for each pipeline which means that it needs the handle

rrawther commented on rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py:
```diff
@@ -21,22 +21,21 @@ def next(self):
                 print(size)
-                self.array = np.array(["                 "])
-                
+                self.array = np.empty([1, size], dtype="<U15")
```
what is dtype "<U15"?

rrawther commented on rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py:
```diff
@@ -83,25 +113,20 @@ def next(self):
 
     def __next__(self):
         if(b.isEmpty(self.loader._handle)):
-            timing_info = b.getTimingInfo(self.loader._handle)
```
we might need to add this back, but make it optional based on DBGLOG etc

rrawther commented on rocAL/rocAL_pybind/example/new_api/READERS_TEST_FILE.sh:
```diff
@@ -0,0 +1,296 @@
+if [[ $# -gt 0 ]]; then
```
please add shellbang before this line

swetha097 replied:
Done

rrawther commented on rocAL/rocAL_pybind/example/new_api/rocAL_api_coco_pipeline.py:
```diff
@@ -0,0 +1,337 @@
+          self.encoded_labels = torch.as_tensor(labels_array, dtype=torch.int32, device=torch_gpu_device)
+          encoded_bboxes_tensor = self.encoded_bboxes.cpu()
+          encodded_labels_tensor = self.encoded_labels.cpu()
+          #encoded_bboxes_tensor = torch.tensor(self.encoded_bboxes.cpu).view(self.bs, -1, 4).contiguous()
```
remove commented lines

swetha097 replied:
Done

rrawther commented on rocAL/rocAL_pybind/example/new_api/rocAL_api_tf_detection_pipeline.py:
```diff
@@ -0,0 +1,132 @@
+import os
+from parse_config import parse_args
+
+
```
remove empty lines

swetha097 replied:
Done

rrawther commented on rocAL/rocAL_pybind/example/tf_mnistTrainingExample/README.md:
```diff
@@ -1,35 +0,0 @@
-## Tensorflow MNIST digit classification example with rocAL pipeline
```
please add the readme back

swetha097 replied:
The mnist example will be added as a part of mini trainings PR.

rrawther commented on rocAL/rocAL_pybind/rali_pybind.cpp:
```diff
@@ -95,10 +95,43 @@ namespace rali{
+
+    py::object wrapper_label_copy(RaliContext context, py::object p)
+    {
+        // auto buf = array.request();
```
remove commented lines. applies to others functions as well

swetha097 replied:
Done

rrawther reviewed (changes requested):
@swetha097 Please address the review comments

LakshmiKumar23 commented:
@kiritigowda Please wait to merge this

rrawther commented on rocAL/rocAL_pybind/amd/rocal/decoders.py:
```diff
@@ -0,0 +1,264 @@
+        return (decoded_image)
+
+
+def image_random_crop(*inputs,user_feature_key_map=None ,path = '', file_root= '', annotations_file='', num_shards = 1, shard_id = 0, random_shuffle = False, affine=True, bytes_per_sample_hint=0, device_memory_padding= 16777216, host_memory_padding = 8388608, hybrid_huffman_threshold = 1000000,
```
please remove unused arguments like device_memory_padding etc from this definition

rrawther reviewed (changes requested):
@swetha097 please address review comments

rrawther commented on rocAL/rocAL_pybind/amd/rocal/decoders.py:
```diff
@@ -0,0 +1,264 @@
+import amd.rocal.types as types
```
I am trying this PR. Looks like ops.py is deleted from amd/rali and not added back in amd/rocal.
Please fix that

rrawther reviewed (approved):
@swetha097 some review comments are still not addressed in the PR. Please fix it in future PR

LakshmiKumar23 commented:
@rrawther @swetha097 Opened #813 for remaining issues. This PR can be merged

rrawther commented:
@kiritigowda : Can we merge this PR?
